### PR TITLE
Only annotate forms when linting in the compiler

### PR DESCRIPTION
**Pull request by josevalim**
A lot of time spent on linting is due to eval_file_attribute/2
function that recursively traverses all the AST, annotating
the file name on the AST nodes. This traversal happens so
it is easier to add errors and warnings later on by simply
introspecting the node.

This patch changes eval_file_attribute/2 to only annotate
forms (i.e. attributes and functions) and rely on the #lint
record to keep the proper file information (which it already
did before this patch).

To summarize, both pre scan and pos scan will use the annotate
file attribute in forms, however form/2 already maintains the
proper file in #lint, which we pass around when retrieving the
location information.

After this patch, linting of a regular erlang module with 500LOC
became twice faster when measured with eprof.

### CODE REVIEW

**Comment by josevalim** [comment]
Here is a gist with eprof results: https://gist.github.com/josevalim/59ac9dd7689d256094a1

This patch along side #728 and #727 brings compilation in Erlang 18.0-rc2 from 10% slower to 10% faster when compared to Erlang 17.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by proxyles** [comment]
This patch fails yecc_SUITE:otp_7292 and yecc_SUITE:syntax
Looks like it points towards .yrl and not the .erl

**Comment by josevalim** [comment]
Great catch. I missed the annotation of called functions, which are analyzed in the post scan. I have fixed the regression (one line fix), gave erl_lint another double check and pushed (with --force).

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by josevalim** [comment]
Thanks @proxyles for merging. I have found a similar regression to the called functions one but this time related to types and I have sent a fix on #747.
